File: app/src/main/java/gc/david/dfm/elevation/ElevationInteractor.java
package gc.david.dfm.elevation;

import com.google.android.gms.maps.model.LatLng;
import com.google.common.base.Function;
import com.google.common.collect.Lists;

import java.util.List;

import gc.david.dfm.elevation.model.ElevationModel;
import gc.david.dfm.elevation.model.Result;
import gc.david.dfm.executor.Executor;
import gc.david.dfm.executor.Interactor;
import gc.david.dfm.executor.MainThread;

/**
 * Created by david on 05.01.17.
 */
public class ElevationInteractor implements Interactor, ElevationUseCase {

    private final Executor            executor;
    private final MainThread          mainThread;
    private final ElevationRepository repository;

    private Callback     callback;
    private List<LatLng> coordinateList;

    public ElevationInteractor(final Executor executor,
                               final MainThread mainThread,
                               final ElevationRepository repository) {
        this.executor = executor;
        this.mainThread = mainThread;
        this.repository = repository;
    }

    @Override
    public void execute(final List<LatLng> coordinateList, final Callback callback) {
        if (callback == null) {
            throw new IllegalArgumentException("Callback can't be null, the client of this interactor needs to get the response in the callback");
        }
        this.coordinateList = coordinateList;
        this.callback = callback;
        this.executor.run(this);
    }

    @Override
    public void run() {
        final String coordinatesPath = getCoordinatesPath(coordinateList);

        try {
            final ElevationModel elevationModel = repository.getElevation(coordinatesPath);

            if ("OK".equals(elevationModel.getStatus())) {
                final List<Double> elevationList = getElevationListFromModel(elevationModel);

                mainThread.post(new Runnable() {
                    @Override
                    public void run() {
                        callback.onElevationLoaded(elevationList);
                    }
                });
            } else {
                mainThread.post(new Runnable() {
                    @Override
                    public void run() {
                        callback.onError();
                    }
                });
            }
        } catch (Exception exception) {
            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    callback.onError();
                }
            });
        }
    }

    private String getCoordinatesPath(final List<LatLng> coordinateList) {
        String positionListUrlParameter = "";
        for (int i = 0; i < coordinateList.size(); i++) {
            final LatLng coordinate = coordinateList.get(i);
            positionListUrlParameter += String.valueOf(coordinate.latitude) +
                                        "," +
                                        String.valueOf(coordinate.longitude);
            if (i != coordinateList.size() - 1) {
                positionListUrlParameter += "|";
            }
        }
        return positionListUrlParameter;
    }

    private List<Double> getElevationListFromModel(final ElevationModel elevationModel) {
        return Lists.transform(elevationModel.getResults(), new Function<Result, Double>() {
            @Override
            public Double apply(final Result result) {
                return result.getElevation();
            }
        });
    }
}

File: app/src/main/java/gc/david/dfm/elevation/domain/ElevationInteractor.kt
/*
 * Copyright (c) 2019 David Aguiar Gonzalez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gc.david.dfm.elevation.domain

import com.google.android.gms.maps.model.LatLng
import gc.david.dfm.elevation.data.ElevationRepository
import gc.david.dfm.elevation.data.mapper.ElevationEntityDataMapper
import gc.david.dfm.elevation.data.model.ElevationEntity
import gc.david.dfm.elevation.data.model.ElevationStatus
import gc.david.dfm.elevation.domain.ElevationInteractor.Params
import gc.david.dfm.elevation.domain.model.Elevation
import gc.david.dfm.executor.CoInteractor
import gc.david.dfm.executor.Either
import gc.david.dfm.executor.Failure
import javax.inject.Inject

/**
 * Created by david on 05.01.17.
 */
class ElevationInteractor @Inject constructor(
        private val mapper: ElevationEntityDataMapper,
        private val repository: ElevationRepository
) : CoInteractor<Elevation, Params>() {

    override suspend fun run(params: Params): Either<Failure, Elevation> {
        if (params.coordinateList.isEmpty()) {
            return Either.Left(ElevationFailure.EmptyList())
        } else {
            val coordinatesPath = getCoordinatesPath(params.coordinateList)

            repository.getElevation("", 1).
            return repository.getElevation(coordinatesPath, params.maxSamples)
                    .either({
                        Either.Left(Failure.ServerError)
                    }, {
                        if (ElevationStatus.OK == it.status) {
                            val elevation = mapper.transform(it)
                            Either.Right(elevation)
                        } else {
                            Either.Left(it.status.toString())
                        }
                    })
        }
    }

    private fun getCoordinatesPath(coordinateList: List<LatLng>): String {
        return coordinateList.joinToString("|") { "${it.latitude},${it.longitude}" }
    }

    data class Params(val coordinateList: List<LatLng>, val maxSamples: Int)

    sealed class ElevationFailure : Failure.FeatureFailure() {

        class EmptyList: FeatureFailure()
    }
}
